File: dexfile/src/main/java/com/github/netomi/bat/dexfile/io/ByteBufferBackedDexDataOutput.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.io;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.channels.Channels;
import java.nio.channels.WritableByteChannel;

/**
 * @author Thomas Neidhart
 */
public class ByteBufferBackedDexDataOutput
implements   DexDataOutput
{
    private ByteBuffer byteBuffer;
    private int lastMemberIndex;

    public ByteBufferBackedDexDataOutput(int bytes) {
        byteBuffer = ByteBuffer.allocate(bytes);
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN);
    }

    @Override
    public ByteOrder order() {
        return byteBuffer.order();
    }

    @Override
    public void order(ByteOrder byteOrder) {
        byteBuffer.order(byteOrder);
    }

    @Override
    public int getOffset() {
        return byteBuffer.position();
    }

    @Override
    public void setOffset(int offset) {
        byteBuffer.position(offset);
    }

    @Override
    public void writePadding(int bytes) {
        for (int i = 0; i < bytes; i++) {
            writeByte((byte) 0);
        }
    }

    @Override
    public void writeAlignmentPadding(int alignment) {
        int currentAlignment = getOffset() % alignment;
        int padding = (alignment - currentAlignment) % alignment;
        for (int i = 0; i < padding; i++) {
            byteBuffer.put((byte) 0x0);
        }
    }

    @Override
    public void writeBytes(byte[] bytes) {
        byteBuffer.put(bytes);
    }

    @Override
    public void writeByte(byte b) {
        byteBuffer.put(b);
    }

    @Override
    public void writeUnsignedByte(short b) {
    }

    @Override
    public void writeShort(short value) {
        byteBuffer.putShort(value);
    }

    @Override
    public void writeShort(short value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeUnsignedShort(int value) {
        byteBuffer.putShort((short) value);
    }

    @Override
    public void writeChar(char value) {
        byteBuffer.putChar(value);
    }

    @Override
    public void writeChar(char value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeInt(int value) {
        byteBuffer.putInt(value);
    }

    @Override
    public void writeInt(int value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeUnsignedInt(long value) {
        byteBuffer.putInt((int) value);
    }

    @Override
    public void writeUnsignedInt(int value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeLong(long value) {
        byteBuffer.putLong(value);
    }

    @Override
    public void writeLong(long value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeFloat(float value) {
        byteBuffer.putFloat(value);
    }

    @Override
    public void writeFloat(float value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeDouble(double value) {
        byteBuffer.putDouble(value);
    }

    @Override
    public void writeDouble(double value, int bytes) {
        writePadding(bytes);
    }

    @Override
    public void writeUleb128(int value) {
        do {
            byte b = (byte) (value & 0x7f);
            value >>>= 7;
            if (value != 0) {
                b |= 0x80;
            }
            byteBuffer.put(b);
        } while (value != 0);
    }

    @Override
    public void writeUleb128p1(int value) {
        writeUleb128(value + 1);
    }

    @Override
    public void writeSleb128(int value) {
        while (true) {
            byte b = (byte) (value & 0x7f);
            value >>= 7;
            if (value ==  0 && ((b & 0x40) == 0) ||
                value == -1 && ((b & 0x40) != 0)) {
                byteBuffer.put(b);
                break;
            }

            byteBuffer.put((byte) (b | 0x80));
        }
    }

    public void copy(OutputStream output) throws IOException {
        byteBuffer.position(0);
        WritableByteChannel channel = Channels.newChannel(output);
        channel.write(byteBuffer);
    }
}

File: dexfile/src/main/kotlin/com/github/netomi/bat/dexfile/io/ByteBufferBackedDexDataOutput.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.io

import java.io.IOException
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.Channels

class ByteBufferBackedDexDataOutput(bytes: Int) : DexDataOutput {
    internal val byteBuffer: ByteBuffer

    init {
        byteBuffer = ByteBuffer.allocate(bytes)
        byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
    }

    override var order: ByteOrder
        get() = byteBuffer.order()
        set(value) {
            byteBuffer.order(value)
        }

    override var offset: Int
        get() = byteBuffer.position()
        set(offset) {
            byteBuffer.position(offset)
        }

    override fun writePadding(bytes: Int) {
        for (i in 0 until bytes) {
            writeByte(0)
        }
    }

    override fun writeAlignmentPadding(alignment: Int) {
        val currentAlignment = offset % alignment
        val padding = (alignment - currentAlignment) % alignment
        for (i in 0 until padding) {
            writeByte(0)
        }
    }

    override fun writeBytes(bytes: ByteArray) {
        byteBuffer.put(bytes)
    }

    override fun writeByte(b: Byte) {
        byteBuffer.put(b)
    }

    override fun writeUnsignedByte(b: Short) {
        writeByte(b.toByte())
    }

    override fun writeShort(value: Short) {
        byteBuffer.putShort(value)
    }

    override fun writeShort(value: Short, bytes: Int) {
        require(bytes in 1 .. 2)
        writeInt(value.toInt(), bytes)
    }

    override fun writeUnsignedShort(value: Int) {
        writeShort(value.toShort())
    }

    override fun writeChar(value: Char, bytes: Int) {
        require(bytes in 1 .. 2)
        writeInt(value.code, bytes)
    }

    override fun writeInt(value: Int) {
        byteBuffer.putInt(value)
    }

    override fun writeInt(value: Int, bytes: Int) {
        require(bytes in 1 .. 4)
        var tmpValue = value
        for (i in 0 until bytes) {
            writeByte(tmpValue.toByte())
            tmpValue = tmpValue ushr 8
        }
    }

    override fun writeUnsignedInt(value: Long) {
        writeInt(value.toInt())
    }

    override fun writeLong(value: Long, bytes: Int) {
        require(bytes in 1 .. 8)
        var tmpValue = value
        for (i in 0 until bytes) {
            writeByte(tmpValue.toByte())
            tmpValue = tmpValue ushr 8
        }
    }

    override fun writeFloat(value: Float, bytes: Int) {
        require(bytes in 1 .. 4)
        var bits = java.lang.Float.floatToIntBits(value)
        bits = bits ushr (32 - (8 * bytes))
        for (i in 0 until bytes) {
            writeByte(bits.toByte())
            bits = bits ushr 8
        }
    }

    override fun writeDouble(value: Double, bytes: Int) {
        require(bytes in 1 .. 8)
        var bits = java.lang.Double.doubleToLongBits(value)
        bits = bits ushr (64 - (8 * bytes))
        for (i in 0 until bytes) {
            writeByte(bits.toByte())
            bits = bits ushr 8
        }
    }

    override fun writeUleb128(value: Int) {
        var count    = 0
        var tmpValue = value
        do {
            check(++count <= 5)
            var b    = tmpValue and 0x7f
            tmpValue = tmpValue ushr 7
            if (tmpValue != 0) {
                b = b or 0x80
            }
            writeByte(b.toByte())
        } while (tmpValue != 0)
    }

    override fun writeUleb128p1(value: Int) {
        writeUleb128(value + 1)
    }

    override fun writeSleb128(value: Int) {
        var count    = 0
        var tmpValue = value
        while (true) {
            check(++count <= 5)
            val b    = tmpValue and 0x7f
            tmpValue = tmpValue shr 7
            if (tmpValue ==  0 && (b and 0x40) == 0 ||
                tmpValue == -1 && (b and 0x40) != 0) {
                writeByte(b.toByte())
                break
            }
            writeByte((b or 0x80).toByte())
        }
    }

    @Throws(IOException::class)
    fun copy(output: OutputStream) {
        byteBuffer.position(0)
        val channel = Channels.newChannel(output)
        channel.write(byteBuffer)
    }

    fun toArray(): ByteArray {
        val size  = byteBuffer.position()
        val array = ByteArray(size)
        byteBuffer.position(0)
        byteBuffer.get(array)
        return array
    }
}